swift/apple/FirezoneNetworkExtension/Adapter.swift

```diff
@@ -313,71 +226,109 @@ extension Adapter {
+        if packetTunnelProvider?.reasserting == false {
+          // Tell the UI we're not connected
+          packetTunnelProvider?.reasserting = true
+        }
+      } else {
+        self.logger.log("\(#function): Detected network change: Online.")
+
+        // Hint to connlib we're back online, but only do so if our primary interface changes,
+        // and therefore we need to bump sockets. On darwin, this is needed to send packets
+        // out of a different interface even when 0.0.0.0 is used as the source.
+        // If our primary interface changes, we can be certain the old socket shouldn't be
+        // used anymore.
+        if (path.availableInterfaces.first?.name != primaryInterfaceName) {
+          session.reconnect()
+          primaryInterfaceName = path.availableInterfaces.first?.name
+        }
 
-    case .stoppedTunnelTemporarily:
-      guard path.status == .satisfied else { return }
+        if shouldFetchSystemResolvers(path: path) {
+          // Spawn a new thread to avoid blocking the UI on iOS
+          Task {
+            let resolvers = getSystemDefaultResolvers(interfaceName: path.availableInterfaces.first?.name)
 
-      self.logger.log("Adapter.didReceivePathUpdate: Back online. Starting connlib.")
+            if lastFetchedResolvers.sorted() != resolvers.sorted() {
+              session.setDns(
+                try! String(
+                  decoding: JSONEncoder().encode(resolvers),
+                  as: UTF8.self
+                ).intoRustString()
+              )
 
-      do {
-        self.state = .startingTunnel(
-          session: try WrappedSession.connect(
-            controlPlaneURLString,
-            token,
-            self.getOrCreateFirezoneId(from: self.firezoneIdFileURL),
-            self.getDeviceName(),
-            self.getOSVersion(),
-            self.connlibLogFolderPath,
-            self.logFilter,
-            self.callbackHandler
-          ),
-          onStarted: { error in
-            if let error = error {
-              self.logger.error(
-                "Adapter.didReceivePathUpdate: Error starting connlib: \(error)")
-              self.packetTunnelProvider?.cancelTunnelWithError(error)
-            } else {
-              self.packetTunnelProvider?.reasserting = false
+              lastFetchedResolvers = resolvers
             }
           }
-        )
-      } catch let error as AdapterError {
-        self.logger.error("Adapter.didReceivePathUpdate: Error: \(error)")
-      } catch {
-        self.logger.error(
-          "Adapter.didReceivePathUpdate: Unknown error: \(error) (fatal)")
-      }
+        }
 
-    case .stoppingTunnel, .stoppedTunnel:
-      // no-op
-      break
+        if packetTunnelProvider?.reasserting == true {
+          packetTunnelProvider?.reasserting = false
+        }
+      }
     }
   }
+
+  #if os(iOS)
+    private func shouldFetchSystemResolvers(path: Network.NWPath) -> Bool {
```

Comment:
Caveat, we only fetch new system resolvers if the available gateways for this path has changed. It's not perfect but gets the job done.